### PR TITLE
luminous: rbd: improve 'import-diff' corrupt input error messages

### PR DESCRIPTION
**Pull request by pdvian**
http://tracker.ceph.com/issues/23038
http://tracker.ceph.com/issues/23607

### CODE REVIEW

**Comment by pdvian** [comment]
@dillaman @smithfarm need review

**Comment by smithfarm** [comment]
@pdvian This appears to be adding a completely new function that wasn't there before?

**Comment by smithfarm** [comment]
@dillaman Thanks.

@pdvian Let's use this PR to do both related backports, as @dillaman suggested. I updated the PR description accordingly.

File: src/tools/rbd/action/Import.cc [hunk 1]
```diff
@@ -576,6 +597,48 @@ static int decode_and_set_image_option(int fd, uint64_t imageopt, librbd::ImageO
...
+  r = utils::read_string(fd, length, &value);
+  if (r < 0) {
+    std::cerr << "rbd: failed to decode metadata value" << std::endl;
+    return r;
+  }
+
+  (*imagemetas)[key] = value;
+  return 0;
+}
+
```
Comment (smithfarm) [inline src/tools/rbd/action/Import.cc]: If these functions are not present in luminous, the changes to these functions are probably irrelevant and you can just omit them.

In other words, I would drop these two functions from the cherry-pick and change the conflicts line to "do_import_metadata and decode_imagemeta not present in luminous".

**Comment by smithfarm** [comment]
Those two new functions are from https://github.com/ceph/ceph/commit/2f0b233a31e - @dillaman is that a candidate for backport?

**Comment by dillaman** [comment]
@smithfarm @pdvian Sure -- pull [1] if it makes it easier and cleaner to backport. 

[1] https://github.com/ceph/ceph/pull/17134

**Comment by theanalyst** [comment]
so do we add https://github.com/ceph/ceph/pull/17134 onto this then?

**Comment by pdvian** [comment]
@theanalyst @dillaman Let me try to cherry-pick #17134 and see if we can get cleaner backport PR.

**Comment by pdvian** [comment]
@dillaman @smithfarm @theanalyst I have cherry-picked #17134 first, then changes related to tracker #23038 and later #23607 changes. It is almost a cleaner backport except for #23038 there was a minor conflict for encode call. Rest looks good. Kindly review.

**Comment by smithfarm** [comment]
the make check failure is a known transient one due to the use of EXPECT_DEATH in `src/test/objectstore/store_test.cc` which is known to be unreliable in multithreaded environments

**Comment by yuriw** [comment]
[wip-yuri2-testing-2018-06-06-1208-luminous](https://trello.com/c/wlCO4DxX/321-wip-yuri2-testing-2018-06-06-1208-luminous)

**Pull request merged**